src/ORM/Search/SearchContext.php

```diff
@@ -176,40 +205,117 @@ public function getQuery($searchParams, $sort = false, $limit = false, $existing
         }
 
-        /** @var DataList $query */
-        $query = $query->sort($sort);
-        $this->setSearchParams($searchParams);
+        return $query->sort($sort);
+    }
 
-        $modelObj = Injector::inst()->create($this->modelClass);
-        $searchableFields = $modelObj->searchableFields();
-        foreach ($this->searchParams as $key => $value) {
-            $key = str_replace('__', '.', $key ?? '');
-            if ($filter = $this->getFilter($key)) {
-                $filter->setModel($this->modelClass);
-                $filter->setValue($value);
-                if (!$filter->isEmpty()) {
-                    if (isset($searchableFields[$key]['match_any'])) {
-                        $searchFields = $searchableFields[$key]['match_any'];
-                        $filterClass = get_class($filter);
-                        $modifiers = $filter->getModifiers();
-                        $query = $query->alterDataQuery(function (DataQuery $dataQuery) use ($searchFields, $filterClass, $modifiers, $value) {
-                            $subGroup = $dataQuery->disjunctiveGroup();
-                            foreach ($searchFields as $matchField) {
-                                /** @var SearchFilter $filterClass */
-                                $filter = new $filterClass($matchField, $value, $modifiers);
-                                $filter->apply($subGroup);
-                            }
-                        });
-                    } else {
-                        $query = $query->alterDataQuery([$filter, 'apply']);
+    /**
+     * Use the global primary search for searching across multiple fields
+     *
+     * @param string|array $searchPhrase
+     */
+    protected function primaryFieldSearch(DataList $query, array $searchableFields, $searchPhrase): DataList
+    {
+        $context = $this;
+        $modelClass = $this->modelClass;
+        return $query->alterDataQuery(function (DataQuery $dataQuery) use ($context, $modelClass, $searchableFields, $searchPhrase) {
+            $formFields = $context->getSearchFields();
+            // Function which takes a search phrase or search term and searches for it across all searchable fields.
+            $searchAcrossFields = function ($searchPhrase, $subGroup) use ($context, $modelClass, $formFields, $searchableFields) {
+                foreach ($searchableFields as $field => $spec) {
+                    $formFieldName = str_replace('.', '__', $field);
+                    $filter = $context->getPrimarySearchFilter($modelClass, $field);
+                    // Only apply filter if the field is allowed to be primary and is backed by a form field.
+                    // Otherwise we could be dealing with, for example, a DataObject which implements scaffoldSearchField
+                    // to provide some unexpected field name, where the below would result in a DatabaseException.
+                    if ((!isset($spec['primary']) || $spec['primary'])
```

Comment:
This will be true if `$spec['primary']` is not set (implicitly allowed) or if `$spec['primary'] == true` (explicitly allowed).
It will be false if `$spec['primary'] === false` for example which is the documented way to disallow a field to be used in the primary field search.

Reply:
done